### DISCUSSION TITLE
How to build [dynamic].js with NextJs getStaticProps

### DISCUSSION THREAD

**Discussion by SergejHb** [vercel/next.js#26052] [Help]
Hello everyone,

I want to generate thousands of product for a catalog which should be static without back-end.

I just start learning NextJs and have found How To instructions on following page:

https://developer.mongodb.com/how-to/nextjs-with-mongodb/

I am interested in “Example 3” - static site generation.

I could connect to the mongodb and fetch data for the top.js (as in example 3), everything works fine.

But I want also to generate every “movie”-page, but got stuck on it already for 2 days.

I have created a additional file [mid].js, here is my current code (after 20+ times rewritten)

Maybe some has an idea what to correct or can give a code example if its completely wrong.
Thank in advance

```
 import { connectToDatabase } from "../util/mongodb";
import { Fragment } from "react";

function MovieDetailPage(props) {
  const { loadedMovie } = props;

  if (!loadedMovie) {
    return <p>Loading...</p>;
  }

  return (
    <Fragment>
      <h1>{loadedMovie.title}</h1>
      
    </Fragment>
  );
}

export async function getStaticProps(context) {
  const { params } = context;

  const movieId = params.mid;

  const data = await connectToDatabase();

  const movie = data.movies.find((movie) => movie.id === movieId);

  if (!movie) {
    return { notFound: true };
  }

  return {
    props: {
      loadedMovie: movie,
    },
  };
}

export async function getStaticPaths() {
  const data = await connectToDatabase();

  const ids = data.movies.map((movie) => movie.id);
  const pathsWithParams = ids.map((id) => ({ params: { mid: id } }));

  return {
    paths: pathsWithParams,
    fallback: true,
  };
}

export default MovieDetailPage;
```

**Comment by akellbl4** [comment]
Hi @SergejHb, I'm ready to help.
What kind of error you are getting or what's wrong? As I can see your example looks pretty good. Maybe I skipped something but it should work.

**Reply by SergejHb** [reply]
Hello @akellbl4 , thank you for fast reply. 
Please find down the error and steps which I did after. 

Maybe you have idea.


The error is 

__Server Error
TypeError: Cannot read property 'map' of undefined_


showing on the line 
 _const ids = data.movies.map((movie) => movie.id);_

I have created a sub-folder in pages/movies/movie and put the [mid].js there.

The is no error any longer in the console or on the server by going to the "movie" page
localhost:3000/movies/id1

But there is a 404 page not found. 

Just for understanding
This folder "pages/movies" with file "movies.js" exists and the page "localhost:3000/movies" receiving all movies.
Currently I am receiving here "movie.title" only .
I have remembered that mongodb has "_id" instead of "id" so I could get the titles with "movie._id" as well.

Coming back to "page of 1 Movie" I have tried to change "id" to "_id". In the console is "compiled successfully" 
but if i go to the page and put 
" localhost:3000/movies/_id" but instead of "_id" the real id from the data base, so the url looks like that.
localhost:3000/movies/60c5cf4e9cb2f3adbc1e9d7b
I am still getting 404 error


Currently it looks like that

```
import { connectToDatabase } from ".../../../util/mongodb";
import { Fragment } from "react";

function MovieDetailPage(props) {
  const { loadedMovie } = props;

  if (!loadedMovie) {
    return <p>Loading...</p>;
  }

  return (
    <Fragment>
      <h1>{loadedMovie.title}</h1>
      
    </Fragment>
  );
}

export async function getStaticProps(context) {
  const { params } = context;

  const movieId = params.mid;

const { db } = await connectToDatabase();
  const movie = await db
    .collection("movies")
   .find((movie) => movie._id === movieId);

  if (!movie) {
    return { notFound: true };
  }

  return {
    props: {
      loadedMovie: movie,
    },
  };
}

export async function getStaticPaths() {
  const data = await connectToDatabase();

  const ids = data.movies.map((movie) => movie._id);
  const pathsWithParams = ids.map((id) => ({ params: { mid: id } })); // I was trying to change "id" in this line to "_id" as well

  return {
    paths: pathsWithParams,
    fallback: true,
  };
}

export default MovieDetailPage;
```

**Reply by SergejHb** [reply]
I was having a mistake in sub-folders, now there is everything ok, but the error is still same
_Server Error
TypeError: Cannot read property 'map' of undefined
